File: lib/heat/heat_grid.dart
import 'dart:math' as math;

import 'package:speedster/domain/track_point.dart';
import 'package:speedster/stats/stats_engine.dart';

/// Eine quantisierte Rasterzelle auf einer Pyramidenebene.
class HeatCell {
  const HeatCell(this.row, this.col);

  final int row;
  final int col;

  @override
  bool operator ==(Object other) =>
      other is HeatCell && other.row == row && other.col == col;

  @override
  int get hashCode => Object.hash(row, col);

  @override
  String toString() => 'HeatCell($row, $col)';
}

/// Ungerichtete Kante zwischen zwei benachbarten Zellen.
class HeatEdgeKey {
  const HeatEdgeKey(this.a, this.b);

  /// Sortiert die Endpunkte, damit Hin- und Rueckrichtung denselben
  /// Schluessel ergeben.
  factory HeatEdgeKey.normalized(HeatCell x, HeatCell y) {
    final xFirst = x.row < y.row || (x.row == y.row && x.col <= y.col);
    return xFirst ? HeatEdgeKey(x, y) : HeatEdgeKey(y, x);
  }

  final HeatCell a;
  final HeatCell b;

  @override
  bool operator ==(Object other) =>
      other is HeatEdgeKey && other.a == a && other.b == b;

  @override
  int get hashCode => Object.hash(a, b);

  @override
  String toString() => 'HeatEdgeKey($a -> $b)';
}

/// Laufende Summen fuer den Schwerpunkt einer Zelle.
class CellAccum {
  double latSum = 0;
  double lngSum = 0;
  int n = 0;

  void add(double lat, double lng) {
    latSum += lat;
    lngSum += lng;
    n++;
  }
}

/// Aggregat einer Fahrt auf einer Rasterebene.
class LevelFold {
  final Map<HeatCell, CellAccum> cells = {};
  final Map<HeatEdgeKey, int> edges = {};
}

/// Aggregat einer Fahrt ueber alle Rasterebenen.
class TripFold {
  TripFold() : levels = List.generate(HeatGrid.levelCount, (_) => LevelFold());

  final List<LevelFold> levels;
}

/// Reine Rasterlogik. Spiegel von `backend/app/Services/HeatGrid.php` —
/// jede Aenderung hier muss dort nachgezogen werden, sonst weicht die
/// Cloud-Heatmap sichtbar von der lokalen ab.
class HeatGrid {
  const HeatGrid._();

  static const List<double> cellMeters = [25.0, 100.0, 400.0];
  static const int levelCount = 3;
  static const double metersPerDegLat = 111320.0;
  static const double cosClamp = 0.01;
  static const double maxAccuracyMeters = 50.0;
  static const double maxGapMeters = 200.0;
  static const int maxGapSeconds = 30;
  static const double resampleMeters = 10.0;

  static double latStep(int level) => cellMeters[level] / metersPerDegLat;

  /// Quantisiert eine Koordinate. [HeatCell.row] haengt nur von [lat] ab und
  /// die Laengengrad-Schrittweite nur von der Zeile — dadurch ist die
  /// Rechnung nicht selbstbezueglich und in jeder Sprache reproduzierbar.
  static HeatCell cellFor(double lat, double lng, int level) {
    final step = latStep(level);
    final row = (lat / step).floor();
    final rowLat = (row + 0.5) * step;
    final lngStep =
        step / math.max(math.cos(rowLat * math.pi / 180.0), cosClamp);
    final col = (lng / lngStep).floor();
    return HeatCell(row, col);
  }

  /// Faltet eine Fahrt in Zell- und Kantenaggregate, fuer alle Level.
  static TripFold foldTrip(List<TrackPoint> points) {
    final fold = TripFold();
    for (final segment in _segments(points)) {
      for (var level = 0; level < levelCount; level++) {
        _foldSegment(segment, level, fold.levels[level]);
      }
    }
    return fold;
  }

  /// Wirft ungenaue Punkte weg und trennt die Spur an Mess-Luecken.
  /// Ohne die Trennung zieht ein GPS-Ausfall im Tunnel eine Gerade quer
  /// durch die Stadt, die sich bei jeder Fahrt aufsummieren wuerde.
  static List<List<TrackPoint>> _segments(List<TrackPoint> points) {
    final result = <List<TrackPoint>>[];
    var current = <TrackPoint>[];

    for (final p in points) {
      if (p.accuracy > maxAccuracyMeters) continue;
      if (current.isEmpty) {
        current.add(p);
        continue;
      }
      final prev = current.last;
      final gap =
          StatsEngine.haversineMeters(prev.lat, prev.lng, p.lat, p.lng);
      final seconds = p.timestamp.difference(prev.timestamp).inSeconds.abs();
      if (gap > maxGapMeters || seconds > maxGapSeconds) {
        if (current.length > 1) result.add(current);
        current = [p];
      } else {
        current.add(p);
      }
    }
    if (current.length > 1) result.add(current);
    return result;
  }

  static void _foldSegment(
    List<TrackPoint> segment,
    int level,
    LevelFold acc,
  ) {
    // Schwerpunkte nur aus echten Messungen.
    for (final p in segment) {
      acc.cells
          .putIfAbsent(cellFor(p.lat, p.lng, level), CellAccum.new)
          .add(p.lat, p.lng);
    }

    final path = <HeatCell>[];
    for (var i = 0; i < segment.length - 1; i++) {
      final from = segment[i];
      final to = segment[i + 1];
      _appendCell(path, cellFor(from.lat, from.lng, level));

      // Nachverdichten, damit bei hohem Tempo keine Zelle uebersprungen wird.
      final distance =
          StatsEngine.haversineMeters(from.lat, from.lng, to.lat, to.lng);
      final steps = (distance / resampleMeters).floor();
      for (var s = 1; s <= steps; s++) {
        final f = s / (steps + 1);
        _appendCell(
          path,
          cellFor(
            from.lat + (to.lat - from.lat) * f,
            from.lng + (to.lng - from.lng) * f,
            level,
          ),
        );
      }
    }
    if (segment.isNotEmpty) {
      final last = segment.last;
      _appendCell(path, cellFor(last.lat, last.lng, level));
    }

    HeatEdgeKey? previous;
    for (var i = 0; i < path.length - 1; i++) {
      final edge = HeatEdgeKey.normalized(path[i], path[i + 1]);
      // Direkt wiederholte Kante = GPS-Zittern ueber die Zellgrenze.
      if (edge == previous) continue;
      acc.edges[edge] = (acc.edges[edge] ?? 0) + 1;
      previous = edge;
    }
  }

  static void _appendCell(List<HeatCell> path, HeatCell cell) {
    if (path.isNotEmpty && path.last == cell) return;
    path.add(cell);
  }
}

File: backend/app/Services/HeatGrid.php
<?php

namespace App\Services;

/**
 * Rasterlogik der Heatmap. Zeilengetreuer Spiegel von
 * `lib/heat/heat_grid.dart` — weicht eine Seite ab, sieht die
 * Cloud-Heatmap anders aus als die lokale. HeatGridParityTest belegt die
 * Gleichheit anhand gemeinsamer Fixtures.
 */
class HeatGrid
{
    public const CELL_METERS = [25.0, 100.0, 400.0];

    public const LEVEL_COUNT = 3;

    public const METERS_PER_DEG_LAT = 111320.0;

    public const COS_CLAMP = 0.01;

    public const MAX_ACCURACY_METERS = 50.0;

    public const MAX_GAP_METERS = 200.0;

    public const MAX_GAP_SECONDS = 30;

    public const RESAMPLE_METERS = 10.0;

    private const EARTH_RADIUS_METERS = 6371000.0;

    public static function latStep(int $level): float
    {
        return self::CELL_METERS[$level] / self::METERS_PER_DEG_LAT;
    }

    /**
     * Quantisiert eine Koordinate.
     *
     * @return array{0:int,1:int} [row, col]
     */
    public static function cellFor(float $lat, float $lng, int $level): array
    {
        $step = self::latStep($level);
        $row = (int) floor($lat / $step);
        $rowLat = ($row + 0.5) * $step;
        $lngStep = $step / max(cos($rowLat * M_PI / 180.0), self::COS_CLAMP);
        $col = (int) floor($lng / $lngStep);

        return [$row, $col];
    }

    public static function haversineMeters(
        float $lat1,
        float $lng1,
        float $lat2,
        float $lng2
    ): float {
        $dLat = deg2rad($lat2 - $lat1);
        $dLng = deg2rad($lng2 - $lng1);
        $a = sin($dLat / 2) ** 2
            + cos(deg2rad($lat1)) * cos(deg2rad($lat2)) * sin($dLng / 2) ** 2;

        return self::EARTH_RADIUS_METERS * 2 * atan2(sqrt($a), sqrt(1 - $a));
    }

    /**
     * @param  array<int, array<string, mixed>>  $points
     * @return array<int, array{cells: array<string, array{lat_sum:float,lng_sum:float,n:int}>, edges: array<string,int>}>
     */
    public static function foldTrip(array $points): array
    {
        $fold = [];
        for ($level = 0; $level < self::LEVEL_COUNT; $level++) {
            $fold[$level] = ['cells' => [], 'edges' => []];
        }

        foreach (self::segments($points) as $segment) {
            for ($level = 0; $level < self::LEVEL_COUNT; $level++) {
                self::foldSegment($segment, $level, $fold[$level]);
            }
        }

        return $fold;
    }

    /**
     * Wirft ungenaue Punkte weg und trennt die Spur an Mess-Luecken. Ohne
     * die Trennung zieht ein GPS-Ausfall im Tunnel eine Gerade quer durch
     * die Stadt, die sich bei jeder Fahrt aufsummieren wuerde.
     *
     * @return array<int, array<int, array<string, mixed>>>
     */
    private static function segments(array $points): array
    {
        $result = [];
        $current = [];

        foreach ($points as $p) {
            if (($p['accuracy'] ?? 0) > self::MAX_ACCURACY_METERS) {
                continue;
            }
            if ($current === []) {
                $current = [$p];

                continue;
            }
            $prev = $current[count($current) - 1];
            $gap = self::haversineMeters(
                $prev['lat'], $prev['lng'], $p['lat'], $p['lng']
            );
            // Altbestand vor dieser Version wurde ohne Zeitstempel abgelegt;
            // dann greift nur die Distanzpruefung.
            $timed = isset($p['t'], $prev['t']);
            $seconds = $timed
                ? abs(strtotime($p['t']) - strtotime($prev['t']))
                : 0;

            if ($gap > self::MAX_GAP_METERS
                || ($timed && $seconds > self::MAX_GAP_SECONDS)) {
                if (count($current) > 1) {
                    $result[] = $current;
                }
                $current = [$p];
            } else {
                $current[] = $p;
            }
        }
        if (count($current) > 1) {
            $result[] = $current;
        }

        return $result;
    }

    private static function foldSegment(array $segment, int $level, array &$acc): void
    {
        // Schwerpunkte nur aus echten Messungen.
        foreach ($segment as $p) {
            [$row, $col] = self::cellFor($p['lat'], $p['lng'], $level);
            $key = "$row:$col";
            if (! isset($acc['cells'][$key])) {
                $acc['cells'][$key] = ['lat_sum' => 0.0, 'lng_sum' => 0.0, 'n' => 0];
            }
            $acc['cells'][$key]['lat_sum'] += $p['lat'];
            $acc['cells'][$key]['lng_sum'] += $p['lng'];
            $acc['cells'][$key]['n']++;
        }

        $path = [];
        $count = count($segment);
        for ($i = 0; $i < $count - 1; $i++) {
            $from = $segment[$i];
            $to = $segment[$i + 1];
            self::appendCell($path, self::cellFor($from['lat'], $from['lng'], $level));

            // Nachverdichten, damit bei hohem Tempo keine Zelle uebersprungen wird.
            $distance = self::haversineMeters(
                $from['lat'], $from['lng'], $to['lat'], $to['lng']
            );
            $steps = (int) floor($distance / self::RESAMPLE_METERS);
            for ($s = 1; $s <= $steps; $s++) {
                $f = $s / ($steps + 1);
                self::appendCell($path, self::cellFor(
                    $from['lat'] + ($to['lat'] - $from['lat']) * $f,
                    $from['lng'] + ($to['lng'] - $from['lng']) * $f,
                    $level
                ));
            }
        }
        if ($count > 0) {
            $last = $segment[$count - 1];
            self::appendCell($path, self::cellFor($last['lat'], $last['lng'], $level));
        }

        $previous = null;
        $pathLength = count($path);
        for ($i = 0; $i < $pathLength - 1; $i++) {
            $edge = self::normalizedEdge($path[$i], $path[$i + 1]);
            // Direkt wiederholte Kante = GPS-Zittern ueber die Zellgrenze.
            if ($edge === $previous) {
                continue;
            }
            $acc['edges'][$edge] = ($acc['edges'][$edge] ?? 0) + 1;
            $previous = $edge;
        }
    }

    private static function appendCell(array &$path, array $cell): void
    {
        $last = $path === [] ? null : $path[count($path) - 1];
        if ($last !== null && $last[0] === $cell[0] && $last[1] === $cell[1]) {
            return;
        }
        $path[] = $cell;
    }

    private static function normalizedEdge(array $x, array $y): string
    {
        $xFirst = $x[0] < $y[0] || ($x[0] === $y[0] && $x[1] <= $y[1]);
        [$a, $b] = $xFirst ? [$x, $y] : [$y, $x];

        return "{$a[0]}:{$a[1]}:{$b[0]}:{$b[1]}";
    }
}
